### PR TITLE
value attribution merge fix and more test cases

### PR DESCRIPTION
**Pull request by aligeti**
What: 
Recent merges dropped the value attribution feature changes.Adding them back with more test cases

Why:
Value attribution changes got dropped in the merge

Please describe the tests:
 - Test 1:
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

File: lib/uplink/bucket_attrs_test.go [hunk 1]
```diff
@@ -53,6 +54,86 @@ func testPlanetWithLibUplink(t *testing.T, cfg testConfig,
 	})
 }
 
+// check that partner bucket attributes are stored and retrieved correctly.
```
Comment (aligeti) [inline lib/uplink/bucket_attrs_test.go]: Rational for code Changes: This test case was dropped during the merge. I added back the change and also improved the test case by making meaningful comparision between expected and actual.

File: lib/uplink/project.go [hunk 2]
```diff
@@ -157,6 +157,20 @@ func (p *Project) OpenBucket(ctx context.Context, bucketName string, access *Enc
 		return nil, err
 	}
```
Comment (aligeti) [inline lib/uplink/project.go]: **Rational for code Changes:** 
This function is updated to handle the addition of 'PartnerID' at the time of bucket creation or at a later time after the bucket was first created and no partnerID was associated with it when it was first created. 

see the above test case **"open with partner id"** and **"open with different partner id"** respectively

File: lib/uplink/project.go [hunk 3]
```diff
@@ -233,3 +247,23 @@ func (p *Project) checkBucketAttribution(ctx context.Context, bucketName string)
 
 	return p.metainfo.SetAttribution(ctx, bucketName, *partnerID)
 }
+
```
Comment (aligeti) [inline lib/uplink/project.go]: **Rational for code Changes:**
This function sets the relevant partner ID information in the protobuf message before sending it to satellite endpoint.

File: pkg/pb/metainfo.pb.go [hunk 4]
```diff
@@ -3,15 +3,17 @@
```
Comment (aligeti) [inline pkg/pb/metainfo.pb.go]: **Rational for code Changes:**
Changes due to protobuf changes and corresponding code generation tool

File: pkg/pb/metainfo.proto [hunk 5]
```diff
@@ -42,12 +42,13 @@ service Metainfo {
 message Bucket {
```
Comment (aligeti) [inline pkg/pb/metainfo.proto]: **Rational for code Changes:**
The changes in this protobuf file is due to the addition of partner_id field

File: satellite/metainfo/metainfo.go [hunk 6]
```diff
@@ -702,19 +702,38 @@ func (endpoint *Endpoint) CreateBucket(ctx context.Context, req *pb.BucketCreate
 		return nil, status.Errorf(codes.InvalidArgument, err.Error())
 	}
```
Comment (aligeti) [inline satellite/metainfo/metainfo.go]: **Rational for code Changes:**
The change here checks if already bucket exists, then update the bucket with the partnerID if one exists, otherwise add the it as a new bucket.

I chose to handle the create/update of bucket info in this same function. Otherwise it needed to add a new function to this satellite endpoint and would call for quite a number of line of new code addition in the satellite and as well as the uplink side.
adding these extra 10-15 lines of code to check if bucket already exists eliminated at least 50-75 lines of new code(from crude estimation).

File: satellite/metainfo/metainfo.go [hunk 7]
```diff
@@ -805,18 +824,25 @@ func (endpoint *Endpoint) SetBucketAttribution(context.Context, *pb.BucketSetAtt
 	return resp, status.Error(codes.Unimplemented, "not implemented")
 }
```
Comment (aligeti) [inline satellite/metainfo/metainfo.go]: **Rational for code Changes:**
added necessary changes to handle conversion from/to proto and storj.bucket

File: satellite/metainfo/metainfo.go [hunk 8]
```diff
@@ -836,9 +862,15 @@ func convertProtoToBucket(req *pb.BucketCreateRequest, projectID uuid.UUID) (sto
 
 func convertBucketToProto(ctx context.Context, bucket storj.Bucket) (pbBucket *pb.Bucket) {
 	rs := bucket.DefaultRedundancyScheme
```
Comment (aligeti) [inline satellite/metainfo/metainfo.go]: **Rational for code Changes:**
added necessary changes to handle conversion from/to proto and storj.bucket

File: satellite/satellitedb/attribution_test.go [hunk 9]
```diff
@@ -87,5 +88,21 @@ func TestUsers(t *testing.T) {
 			DefaultSegmentsSize: int64(100),
 		})
 		require.NoError(t, err)
```
Comment (aligeti) [inline satellite/satellitedb/attribution_test.go]: **Rational for code Changes:**
added test case to test the update bucket db function. Also changed the name from bucket_test to attribution_test.go as these test cases are more related to attribution feature.

File: go.mod [hunk 10]
```diff
@@ -115,7 +115,7 @@ require (
 	go.uber.org/multierr v1.1.0 // indirect
 	go.uber.org/zap v1.10.0
 	golang.org/x/crypto v0.0.0-20190611184440-5c40567a22f8
-	golang.org/x/net v0.0.0-20190628185345-da137c7871d7 // indirect
+	golang.org/x/net v0.0.0-20190628185345-da137c7871d7
```
Comment (aligeti) [inline go.mod]: unable to get past thru linter checker on CI with `// indirect', so had to push without it just to get pass thru build successfully on CI

Comment (coyle) [reply go.mod]: Do you know why it was complaining ?

**Comment by aligeti** [comment]
That I Need to investigate. I had to make that change so that Cl build can be successful. I will look into it.

File: pkg/pb/metainfo.proto [hunk 11]
```diff
@@ -42,12 +42,13 @@ service Metainfo {
...
     bytes                  name = 1;
     encryption.CipherSuite path_cipher = 2;
+    bytes                  partner_id = 3;
```
Comment (JessicaGreben) [inline pkg/pb/metainfo.proto]: before this field was called attribution_id. why the change to partner_id?

File: satellite/metainfo/metainfo.go [hunk 12]
```diff
@@ -674,8 +674,13 @@ func (endpoint *Endpoint) GetBucket(ctx context.Context, req *pb.BucketGetReques
 		return nil, status.Errorf(codes.Internal, err.Error())
 	}
 
+	convBucket, err := convertBucketToProto(ctx, bucket)
```
Comment (JessicaGreben) [inline satellite/metainfo/metainfo.go]: convBucket is a weird name. could you just call it bucket or protoBucket or pbBucket?

File: uplink/metainfo/client.go [hunk 13]
```diff
@@ -316,14 +334,21 @@ func convertBucketToProtoRequest(bucket storj.Bucket) pb.BucketCreateRequest {
...
+	}, nil
 }
 
-func convertProtoToBucket(pbBucket *pb.Bucket) storj.Bucket {
+func convertProtoToBucket(pbBucket *pb.Bucket) (bucket storj.Bucket, err error) {
 	defaultRS := pbBucket.GetDefaultRedundancyScheme()
 	defaultEP := pbBucket.GetDefaultEncryptionParameters()
+	var partnerID uuid.UUID
+	err = partnerID.UnmarshalJSON(pbBucket.GetPartnerId())
+	// partnerID not ie zero, is not error)
```
Comment (JessicaGreben) [inline uplink/metainfo/client.go]: this comment doesn't make sense

File: uplink/metainfo/client.go [hunk 14]
```diff
@@ -298,11 +310,17 @@ func (client *Client) ListBuckets(ctx context.Context, listOpts storj.BucketList
 	return resultBucketList, nil
 }
 
-func convertBucketToProtoRequest(bucket storj.Bucket) pb.BucketCreateRequest {
+func convertBucketToProtoRequest(bucket storj.Bucket) (bucketReq pb.BucketCreateRequest, err error) {
 	rs := bucket.DefaultRedundancyScheme
+	//TODO @ASK handle err
```
Comment (JessicaGreben) [inline uplink/metainfo/client.go]: is this comment still relevant? or should it be removed?

File: uplink/metainfo/client.go [hunk 15]
```diff
@@ -316,14 +334,21 @@ func convertBucketToProtoRequest(bucket storj.Bucket) pb.BucketCreateRequest {
...
-	}
+	}, nil
 }
 
-func convertProtoToBucket(pbBucket *pb.Bucket) storj.Bucket {
+func convertProtoToBucket(pbBucket *pb.Bucket) (bucket storj.Bucket, err error) {
 	defaultRS := pbBucket.GetDefaultRedundancyScheme()
 	defaultEP := pbBucket.GetDefaultEncryptionParameters()
+	var partnerID uuid.UUID
+	err = partnerID.UnmarshalJSON(pbBucket.GetPartnerId())
```
Comment (JessicaGreben) [inline uplink/metainfo/client.go]: many places in the code use this function `bytesToUUID()`. Not sure if that is better than this unmarshalling

File: satellite/satellitedb/buckets.go [hunk 16]
```diff
@@ -73,6 +73,20 @@ func (db *bucketsDB) GetBucket(ctx context.Context, bucketName []byte, projectID
 	return convertDBXtoBucket(dbxBucket)
 }
 
+// UpdateBucket upates a bucket
+func (db *bucketsDB) UpdateBucket(ctx context.Context, bucket storj.Bucket) (_ storj.Bucket, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	var updateFields dbx.BucketMetainfo_Update_Fields
+	updateFields.PartnerId = dbx.BucketMetainfo_PartnerId(bucket.PartnerID[:])
```
[the remark this answers is not in the document]
Comment (JessicaGreben) [reply satellite/satellitedb/buckets.go]: @aligeti if you make this ^ change, would you mind updating the `CreateBucket` function as well?

File: satellite/metainfo/metainfo.go [hunk 17]
```diff
@@ -674,8 +674,13 @@ func (endpoint *Endpoint) GetBucket(ctx context.Context, req *pb.BucketGetReques
...
+	if err != nil {
+		return resp, err
```
Comment (JessicaGreben) [inline satellite/metainfo/metainfo.go]: i think this should return a status error

Comment (aligeti) [reply satellite/metainfo/metainfo.go]: updated

File: satellite/metainfo/metainfo.go (hunk 6 above)
Comment (JessicaGreben) [reply satellite/metainfo/metainfo.go]: i think its better to add a separate `UpdateBucket` function

File: uplink/metainfo/client.go (hunk 14 above)
Comment (aligeti) [reply uplink/metainfo/client.go]: I remember removing it.. apparently not :-).. it should be removed.... removed.

File: satellite/metainfo/metainfo.go (hunk 6 above)
Comment (aligeti) [reply satellite/metainfo/metainfo.go]: I missed that on it and @ethanadams pointed it out. I handled it with 3 lines of code and added a test case as well in TestPartnerBucketAttrs()

File: uplink/metainfo/client.go (hunk 15 above)
Comment (aligeti) [reply uplink/metainfo/client.go]: I totally agree with you on that. I modified it to use uuid package for the files and functions I modified. I see there were 37reference all over satellite. My recommendation, is to do incremental changes as we know it works for now.

File: uplink/metainfo/client.go (hunk 13 above)
Comment (aligeti) [reply uplink/metainfo/client.go]: agree with u..

File: lib/uplink/project.go [hunk 18]
```diff
@@ -233,3 +247,23 @@ func (p *Project) checkBucketAttribution(ctx context.Context, bucketName string)
...
+// updateBucket updates an existing bucket's attribution info.
+func (p *Project) updateBucket(ctx context.Context, bucketInfo storj.Bucket) (bucket storj.Bucket, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	partnerID, err := uuid.Parse(p.uplinkCfg.Volatile.PartnerID)
```
[the remark this answers is not in the document]
Comment (aligeti) [reply lib/uplink/project.go]: updated as suggested. you right  the updateBucket can be used as a helper function.

File: satellite/metainfo/metainfo.go [hunk 19]
```diff
@@ -702,19 +707,55 @@ func (endpoint *Endpoint) CreateBucket(ctx context.Context, req *pb.BucketCreate
...
-	bucket, err = endpoint.metainfo.CreateBucket(ctx, bucket)
-	if err != nil {
-		return nil, Error.Wrap(err)
+		// partnerID not set
+		if partnerID.IsZero() {
+			return resp, status.Errorf(codes.AlreadyExists, "Bucket already exists")
+		}
+
+		//update the bucket
+		bucket.PartnerID = partnerID
```
Comment (JessicaGreben) [inline satellite/metainfo/metainfo.go]: should we check if the bucket already has this partnerID set? for example, if bucket.Partner == partnerID, then return

File: satellite/metainfo/metainfo.go [hunk 20]
```diff
@@ -702,19 +707,55 @@ func (endpoint *Endpoint) CreateBucket(ctx context.Context, req *pb.BucketCreate
 		return nil, status.Errorf(codes.InvalidArgument, err.Error())
 	}
 
-	bucket, err := convertProtoToBucket(req, keyInfo.ProjectID)
-	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, err.Error())
-	}
+	bucket, err := endpoint.metainfo.GetBucket(ctx, req.GetName(), keyInfo.ProjectID)
+	if err == nil {
```
Comment (JessicaGreben) [inline satellite/metainfo/metainfo.go]: if you are going to keep this update functionality in the create method, can you at least add a comment explaining what this block is about

File: satellite/metainfo/metainfo.go (hunk 19 above)
Comment (aligeti) [reply satellite/metainfo/metainfo.go]: There is checkBucketAttribution() before updateBucket checks if that bucket has already an attribution done. If it done, it returns error even before going to next line too updateBucket. an attribution has to exist even before the bucket is created. So it is handled.

File: satellite/satellitedb/buckets.go [hunk 21]
```diff
@@ -172,10 +186,16 @@ func convertDBXtoBucket(dbxBucket *dbx.BucketMetainfo) (bucket storj.Bucket, err
...
+	partnerID, err := bytesToUUID(dbxBucket.PartnerId)
+	if err != nil {
+		return bucket, err
+	}
+
 	return storj.Bucket{
 		ID:                  id,
 		Name:                string(dbxBucket.Name),
 		ProjectID:           project,
+		PartnerID:           partnerID,
```
[the remark this answers is not in the document]
Comment (ethanadams) [reply satellite/satellitedb/buckets.go]: I'm still seeing 00000000000000000000000000000000 stored in partnerID in projects and bucket_metainfo tables.

Do clean storj-sim setup to reproduce.

File: satellite/satellitedb/buckets.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (aligeti) [reply satellite/satellitedb/buckets.go]: Also added the logic to make sure we are not storing zeros as the UUID in DB instead of NULL

File: satellite/metainfo/metainfo.go (hunk 20 above)
Comment (aligeti) [reply satellite/metainfo/metainfo.go]: added the comment briefing the logic

File: lib/uplink/project.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (egonelbre) [reply lib/uplink/project.go]: This shouldn't be an UUID it should be an user agent string.
